_posts/2019-04-15-Keep-your-RxJS-streams-from-bursting-their-banks.md

```diff
@@ -0,0 +1,612 @@
+But what happens when `param$` emits a new value? The `obj`property will be replaced eventually, but the event stream for the first object will still exist.
+So you could start keeping track of subscriptions and cancel them at the ‘right’ moment:
+
+```typescript
+subscription1 = param$.subscribe(param => {
+  if (subscription2 && !subscription2.closed) subscription2.unsubscribe(); 
+  subscription2 = apiService.getObject(param).subscribe(obj => {
+    this.obj = obj;
+    if (subscription3 && !subscription3.closed) subscription3.unsubscribe(); 
+    subscription3 = anotherApiService.getEvents(obj.busId).subscribe(events => {
+      this.events = events;
+    });
+  });
+});
+
+onDestroy() {
+  subscription1.unsubscribe();
+  if (subscription2 && !subscription2.closed) subscription2.unsubscribe();
+  if (subscription3 && !subscription3.closed) subscription3.unsubscribe();
+}
+```
+
+Wow! That’s amazing!
+Except it isn’t.
+It’s even worse!
+Here are 3 subscriptions, while it could be done with just 1 when using the correct operators:
+
+```typescript
+subscription1 = param$.pipe(
+  switchMap(param => apiService.getObject(param)),
+  tap(obj => this.obj = obj),
+  switchMap(obj => anotherApiService.getEvents(obj.busId))
+).subscribe(events => {
+  this.events = events;
+});
+
+onDestroy() {
+  subscription1.unsubscribe();
+}
+```
+
+This way, only 1 subscription will exist, canceling it will automatically cancel any inner streams.
+And when `param$` would emit a new value, the `switchMap` operator will automatically cancel its inner stream and create a new one.
+
+> RULE: NEVER SUBSCRIBE WITHIN A SUBSCRIPTION
+This also applies for calling a function inside a subscription when that function has a subscription.
+
+## Sharing is caring
+
+<p style="text-align: center;">
+  <img alt="Cat with big bowl of food: What do you mean, I have to share?" src="{{ '/img/2019-04-15-Keep-your-RxJS-streams-from-bursting-their-banks/2y7xld.jpg' | prepend: site.baseurl }}">
+</p>
+
+When using a stream on multiple locations, remember that there is a difference between a cold and a hot Observable.
+In short: a cold Observable will restart its stream for each subscription, while a hot Observable will reuse an existing stream when a subscription is added.
+Think of a cold Observable as a HTTP request, which fires each time again, and a hot Observable as a keyPress stream which emits the same event no matter how many subscriptions exist on the stream.
+If you need more details, you can read this article by Ben Lesh [Hot vs Cold Observables](https://medium.com/@benlesh/hot-vs-cold-observables-f8094ed53339).
+
+Sometimes you'll need to react on a cold Observable in multiple separate streams.
+For example, you need to display the response, but you also need to initiate a new stream to get other data.
+So you'll create two subscriptions and use a **mergeMap**, because subscriptions inside subscriptions are not allowed:
```

Comment:
Seems like some code references are styled with code styling (backticks) and others with bold styling. Probably best to pick one of the two instead of mixing.